lib/khttp/ktest/mocks.go

```diff
@@ -0,0 +1,145 @@
+// +build !release
+
+package ktest
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	rpcAstore "github.com/enfabrica/enkit/astore/rpc/astore"
+	"github.com/enfabrica/enkit/astore/server/astore"
+	"github.com/enfabrica/enkit/lib/srand"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+	"log"
+	"math/rand"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+type MinioDescriptor struct {
+	Port uint16
+}
+
+//RunMinioServer will spin up a minio serer using the local docker daemon
+//it also returns a func to call that will close and destroy the running image
+//the port and network bind are determined by docker and returned
+func RunMinioServer() (MinioDescriptor, func() error, error) {
+	_ = "gcloud beta emulators datastore start --no-store-on-disk"
+	return MinioDescriptor{}, nil, nil
+}
+
+type EmulatedDatastoreDescriptor struct {
+	Addr *net.TCPAddr
+}
+
+func RunEmulatedDatastore() (*EmulatedDatastoreDescriptor, KillAbleProcess, error) {
+	emulatorAddr, err := AllocatePort()
+	if err != nil {
+		return nil, nil, err
+	}
+	cmd := exec.Command("gcloud",
+		"beta", "emulators", "datastore", "start",
+		"--no-store-on-disk",
+		fmt.Sprintf("--host-port=127.0.0.1:%d", emulatorAddr.Port),
```

Comment:
In order:

- This works because listeners don't bind to multiple IP addresses. This is listening on ::<port>, while the emulator is binding to 127.0.0.1. The specific ":0" behaviour is in the tcp.Listen() godoc. 
- Listener is closed at the end of the test since it's bound to the go process. 
- In another upstream I have it returning the whole listener with address as an update, but I support I can do it here as well

Reply:
ack.